fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/SumLiteralRewrite.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalAggregate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.thrift.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * sum(expr +/- literal) ==> sum(expr) +/- literal * count(expr)
+ */
+public class SumLiteralRewrite extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalAggregate()
+                .whenNot(agg -> agg.getSourceRepeat().isPresent())
+                .then(agg -> {
+                    Map<NamedExpression, Pair<Expression, Literal>> funcMap = agg.getOutputs().stream()
+                            .map(this::extractSumLiteral)
+                            .filter(Objects::nonNull)
+                            .collect(Collectors.toMap(item -> item.first, item -> item.second));
+                    if (funcMap.isEmpty()) {
+                        return null;
+                    }
+                    return rewriteSumLiteral(agg, funcMap);
+                }).toRule(RuleType.SUM_LITERAL_REWRITE);
+    }
+
+    private Plan rewriteSumLiteral(
+            LogicalAggregate<?> agg, Map<NamedExpression, Pair<Expression, Literal>> sumLiteralMap) {
+        Set<NamedExpression> newAggOutput = new HashSet<>();
+        for (NamedExpression expr : agg.getOutputExpressions()) {
+            if (!sumLiteralMap.containsKey(expr)) {
+                newAggOutput.add(expr);
+            }
+        }
+
+        Map<Expression, Slot> exprToSum = new HashMap<>();
+        Map<Expression, Slot> exprToCount = new HashMap<>();
+
+        Map<AggregateFunction, NamedExpression> existedAggFunc = new HashMap<>();
+        for (NamedExpression e : agg.getOutputExpressions()) {
+            if (e.children().size() == 1 && e.child(0) instanceof AggregateFunction) {
+                existedAggFunc.put((AggregateFunction) e.child(0), e);
+            }
+        }
+
+        Set<Expression> countSumExpr = new HashSet<>();
+        for (Pair<Expression, Literal> pair : sumLiteralMap.values()) {
+            countSumExpr.add(pair.first);
+        }
+
+        for (Expression e : countSumExpr) {
+            NamedExpression namedSum = constructSum(e, existedAggFunc);
+            NamedExpression namedCount = constructCount(e, existedAggFunc);
+            exprToSum.put(e, namedSum.toSlot());
+            exprToCount.put(e, namedCount.toSlot());
+            newAggOutput.add(namedSum);
+            newAggOutput.add(namedCount);
+        }
+
+        LogicalAggregate<?> newAgg = agg.withAggOutput(ImmutableList.copyOf(newAggOutput));
```

Comment:
we should keep group by key at head of output list

Reply:
The order of origin output is not changed. It's ok